### PR TITLE
Account store state rework

### LINKED ISSUE #67: Decouple reconnect error handling
maybe print the error here and then handle it better later? This kind of all catchers can cause debugging to be hard later

_Originally posted by @KillariDev in https://github.com/DarkFlorist/web3-app/pull/61#discussion_r1098390971_

### PR DESCRIPTION
**Pull request by jubalm**
Addresses #64 which segregate address states

This also resolves #67 #65

### CODE REVIEW

File: app/ts/store/account.ts [hunk 1]
```diff
@@ -1,87 +1,86 @@
...
 					assertsExternalProvider(window.ethereum)
 					const provider = new ethers.providers.Web3Provider(window.ethereum)
 					const signer = provider.getSigner()
-					const address = await signer.getAddress()
-					accountStore.value = { isConnected: true, address }
-				} catch {
-					accountStore.value = accountDefaults
+					return await signer.getAddress()
+				} catch (error) {
+					throw new ConnectAttemptError()
```
Comment (KillariDev) [inline app/ts/store/account.ts:38]: we could try to distinguish different errors here and if they don't match the ones we expect, we throw the error at it is

File: app/ts/store/account.ts [hunk 2]
```diff
@@ -1,87 +1,86 @@
...
 
 	const handleAccountChange = (newAccount: string[]) => {
-		if (ethers.utils.isAddress(newAccount[0])) {
-			accountStore.value = { isConnected: true, address: newAccount[0] }
-			return
-		}
-
-		reset()
-		accountStore.value = accountDefaults
+		accountStore.value = ethers.utils.isAddress(newAccount[0]) ? { state: 'connected', address: newAccount[0] } : accountStoreDefaults
```
Comment (KillariDev) [inline app/ts/store/account.ts:53]: the new account list can be empty too

**Review by KillariDev — APPROVED** [review]

File: app/ts/components/ConnectToWallet.tsx [hunk 3]
```diff
@@ -10,46 +8,38 @@ import * as Icon from './Icon/index.js'
...
+			const connectToWallet = accountStore.value.connect
 			return (
 				<Wrapper>
 					<div class='grid grid-cols-[minmax(auto,max-content)_minmax(auto,max-content)_minmax(auto,max-content)] gap-3 items-center justify-center'>
 						<div class='leading-tight text-right'>Get started quickly by connecting your wallet</div>
 						<div class='transition animate-bounce-x'>
 							<Icon.ArrowRight />
 						</div>
-						<Button class='whitespace-nowrap' onClick={() => connectMutation.dispatch()}>
+						<Button class='whitespace-nowrap' onClick={() => connectToWallet()}>
```
[the remark this answers is not in the document]
Comment (jubalm) [reply app/ts/components/ConnectToWallet.tsx:35]: there's an optional parameter on `connectToWallet` that TS will complain if i place it directly on `onClick`.

[the remark this answers is not in the document]
Comment (MicahZoltu) [reply app/ts/components/ConnectToWallet.tsx:35]: If you make `connectToWallet` an overloaded function does that resolve the problem?

[the remark this answers is not in the document]
Comment (jubalm) [reply app/ts/components/ConnectToWallet.tsx:35]: `connectToWallet` is just a reference to a function with signature `(attemptOnly?: boolean) => void` so JSX `onClick` complains incompatible type when I used it as a parameter. I did try changing the reference function to `(attemptOnly?:boolean) => void | (() => void)` (from what i understand about overriding in TS) but didn't like it either.

Using `bind` no longer gets complaints from TS; as in `connect.bind(null, false)` and i can even use it directly without creating the pointer.

File: app/ts/components/ConnectToWallet.tsx [hunk 4]
```diff
@@ -24,15 +24,14 @@ export const ConnectToWallet = () => {
...
-			const connectToWallet = accountStore.value.connect
 			return (
 				<Wrapper>
 					<div class='grid grid-cols-[minmax(auto,max-content)_minmax(auto,max-content)_minmax(auto,max-content)] gap-3 items-center justify-center'>
 						<div class='leading-tight text-right'>Get started quickly by connecting your wallet</div>
 						<div class='transition animate-bounce-x'>
 							<Icon.ArrowRight />
 						</div>
-						<Button class='whitespace-nowrap' onClick={() => connectToWallet()}>
+						<Button class='whitespace-nowrap' onClick={accountStore.value.connect.bind(null, false)}>
```
Comment (MicahZoltu) [inline app/ts/components/ConnectToWallet.tsx:34]: For the record, I liked `() => accountStore.value.connect()` better than the `bind` solution.  However, I'm fine with all of the options, so pick the one you like best.

I think `undefined` for first parameter of `bind` is more correct than `null` though?
```suggestion
						<Button class='whitespace-nowrap' onClick={accountStore.value.connect.bind(undefined, false)}>
```

File: app/ts/store/account.ts (hunk 2 above)
Comment (jubalm) [reply app/ts/store/account.ts:53]: this will be handled in #70

File: app/ts/components/ConnectToWallet.tsx (hunk 4 above)
Comment (jubalm) [reply app/ts/components/ConnectToWallet.tsx:34]: i'm a fan of arrow functions in event handlers as well but for some reason the type for `accoutStore.value` seem to have lost context as a return value `() => accountStore.value.connect()` in this switch statement.
Screenshot 2023-02-13 at 2 43 11 AM

File: app/ts/store/account.ts (hunk 1 above)
Comment (jubalm) [reply app/ts/store/account.ts]: The try-catch here is an attempt to verify if a wallet account is accessible and is expected to withdraw silently if the attempt failed to assume no account is connected yet. nevertheless, `ConnectAttemptError` will now receive the error message

File: app/ts/components/ConnectToWallet.tsx (hunk 4 above)
Comment (MicahZoltu) [reply app/ts/components/ConnectToWallet.tsx]: Ah, yeah.  that makes sense.  This is why you needed the local, that makes sense.

**Pull request merged**